### PR TITLE
Stop trying to "preserve" iptables counters that are always 0

### PR DESCRIPTION
**Pull request by danwinship**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
@thockin had mentioned a while back that he was pretty sure that all iptables counters are always 0 and we're doing pointless busywork pretending otherwise, and this turns out to be mostly correct.

Specifically, the chain-level counters that our code deals with (as opposed to the rule-level counters that we ignore) are only used for the built-in chains (`INPUT`, `PREROUTING`, etc) that have a default policy, and they count the number of packets/bytes that hit that default policy (ie, that fall off the end of the chain rather than matching a terminal rule). For user-created chains, these counters will always be 0, and can't even manually be set to any other value (ie, if you try to set them in the `iptables-restore` input, the values you provide will be ignored). This is true for both iptables-legacy and iptables-nft.

We never pass `iptables-restore` any input that refers to the built-in chains, so all of our efforts to preserve the counters are pointless. And we don't have to worry about "but what if we wanted to change things in the future so that we did refer to those chains?" because there's no way we could refer to the built-in chains in our `iptables-restore` input without creating race conditions that might accidentally overwrite concurrent modifications to those chains by other users. (This is why the code still uses `ipt.EnsureChain`/`ipt.EnsureRule` for some things rather than doing everything with `ipt.Restore`.)

#### Special notes for your reviewer:
(After this PR, the only thing we use the `iptables-save` output for is to figure out which service/endpoint chains we need to delete. We could theoretically figure that out without using `iptables-save`, but that would require us to more-or-less keep a copy of the current known iptables state in memory at all times. And we'd still need the `iptables-save` version at startup time anyway. A better optimization would be to change it to only do stale-chain cleanup periodically (eg, once a minute) rather than on every resync; the stale chains wouldn't have any rules pointing to them, so they wouldn't cause any problems beyond memory usage. https://github.com/kubernetes/kubernetes/pull/110334 does that.)

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig network
/priority important-longterm

### CODE REVIEW

**Comment by aojea** [comment]
I have the feeling that the parsing will be "less solid" now :grimacing:

**Comment by danwinship** [comment]
It seems strange to worry that the new code will be less correct than the old code, given that the old code was not even being tested against valid data...

**Comment by aojea** [comment]
> It seems strange to worry that the new code will be less correct than the old code, given that the old code was not even being tested against valid data...

ah, is a conclusion based on observation, we didn't have any bug with the old code ... at last, not any that we know :)

**Comment by danwinship** [comment]
Discussion from https://bugzilla.redhat.com/show_bug.cgi?id=2058444 which is unfortunately not public:

----

**Dan Winship   2022-05-31 10:45:54 EDT   Comment 45**

Phil, can you confirm the results of these experiments against the actual code to make sure I'm not missing anything subtle?

1. When using either iptables-legacy or iptables-nft, the system does not keep track of chain-level counts for user-created chains
```
    # iptables -t filter -N FOO
    # iptables -t filter -A OUTPUT -j FOO
    # iptables -t filter -A FOO -d 1.2.3.4 -j REJECT

    # curl http://1.2.3.4/

    # iptables-save -t filter -c | grep FOO
    :FOO - [0:0]
    [53:9070] -A OUTPUT -j FOO
    [1:60] -A FOO -d 1.2.3.4/32 -j REJECT --reject-with icmp-port-unreachable

    # printf "*filter\n:FOO - [9:9999]\nCOMMIT\n" | iptables-restore -n -c
    # iptables-save -t filter -c | grep :FOO
    :FOO - [0:0]
```

ie, the _rule_ has non-0 counters, but the _chain_ always shows [0:0].

2. When using iptables-legacy, doing an "`iptables-restore -n`" (with or without "`-c`") will reset the counters of all chains in restored tables, including chains that are not mentioned in the restore:
```
    # iptables-legacy-save -t filter|grep INPUT
    :INPUT ACCEPT [1626:374227]

    # printf "*filter\n:FOO - [0:0]\n-A FOO -j DROP\nCOMMIT\n" | iptables-legacy-restore -n

    # iptables-legacy-save -t filter|grep INPUT
    :INPUT ACCEPT [13:1626]

    # printf "*filter\n:FOO - [0:0]\n-A FOO -j DROP\nCOMMIT\n" | iptables-legacy-restore -n -c

    # iptables-legacy-save -t filter|grep INPUT
    :INPUT ACCEPT [11:1078]
```

3. When using iptables-nft, doing an "`iptables-restore -n`" (with or without "`-c`") WILL NOT reset the counters of chains that are not mentioned in the restore:
```
    # iptables-nft-save -t filter|grep INPUT
    :INPUT ACCEPT [10313:2631314]

    # printf "*filter\n:FOO - [0:0]\n-A FOO -j DROP\nCOMMIT\n" | iptables-nft-restore -n

    # iptables-nft-save -t filter|grep INPUT
    :INPUT ACCEPT [10378:2638441]

    # printf "*filter\n:FOO - [0:0]\n-A FOO -j DROP\nCOMMIT\n" | iptables-nft-restore -n -c

    # iptables-nft-save -t filter|grep INPUT
    :INPUT ACCEPT [10386:2639973]
```

(kube-proxy does "iptables-save" for two reasons: (a) so it can set counters correctly on restore, (b) so it can see if there are stale KUBE chains lying around that need to be deleted. But it seems like (a) actually never really worked, and we don't need to do (b) on every sync...)

----

**Phil Sutter   2022-05-31 16:30:39 EDT   Comment 46**

> Phil, can you confirm the results of these experiments against the actual code to make sure I'm not missing anything subtle?
> 
> 1. When using either iptables-legacy or iptables-nft, the system does not keep track of chain-level counts for user-created chains

There is no such thing as "chain-level counts". In iptables nomenclature, these values are "policy counters" and are therefore relevant only with chains that support a policy - user-defined ones of course do not. The iptables-restore code is clear there: counter values are parsed only if '`-c`' was given and policy string is not '`-`'.

>     # iptables -t filter -N FOO
>     # iptables -t filter -A OUTPUT -j FOO
>     # iptables -t filter -A FOO -d 1.2.3.4 -j REJECT
> 
>     # curl http://1.2.3.4/
> 
>     # iptables-save -t filter -c | grep FOO
>     :FOO - [0:0]
>     [53:9070] -A OUTPUT -j FOO
>     [1:60] -A FOO -d 1.2.3.4/32 -j REJECT --reject-with icmp-port-unreachable
> 
>     # printf "*filter\n:FOO - [9:9999]\nCOMMIT\n" | iptables-restore -n -c
>     # iptables-save -t filter -c | grep :FOO
>     :FOO - [0:0]
> 
> ie, the _rule_ has non-0 counters, but the _chain_ always shows [0:0].

This was true for `OUTPUT` chain as well if you added the rule to it instead: The chain counter covers packets "falling off the end of the chain" only - add a final `ACCEPT` rule and the counter will remain zero.

> 2. When using iptables-legacy, doing an "`iptables-restore -n`" (with or without "`-c`") will reset the counters of all chains in restored tables, including chains that are not mentioned in the restore:

Yes, I can confirm this and it seems to be a bug in libiptc.

> 3. When using iptables-nft, doing an "`iptables-restore -n`" (with or without "`-c`") WILL NOT reset the counters of chains that are not mentioned in the restore:

IIRC, '`iptables-nft-restore -n`' caches only the bare minimum and therefore does not know what chains actually exist in kernel. Therefore it does not mangle them (for whatever reason).

> (kube-proxy does "iptables-save" for two reasons: (a) so it can set counters correctly on restore, (b) so it can see if there are stale KUBE chains lying around that need to be deleted. But it seems like (a) actually never really worked, and we don't need to do (b) on every sync...)

So (a) requires non-flushing restore to leave other chains' counters alone and is broken with iptables-legacy?

----

**Dan Winship   2022-06-01 08:44:09 EDT   Comment 47**

> > (kube-proxy does "iptables-save" for two reasons: (a) so it can set counters correctly on restore, (b) so it can see if there are stale KUBE chains lying around that need to be deleted. But it seems like (a) actually never really worked, and we don't need to do (b) on every sync...)
> 
> So (a) requires non-flushing restore to leave other chains' counters alone and is broken with iptables-legacy?

No, kube-proxy / kubernetes itself doesn't care about the counters. It's just trying to be a good citizen and not break the counters in case the user was using them for something else. At least... I think that's what it's doing? I haven't done the git archaeology to figure out exactly what the original intent of using "`-c`" was.

But regardless of what the intent was, the code is not actually doing anything useful; our iptables-restore input only ever contains user-defined chains, not built-in ones, so we never actually preserve any counters. We just do a bunch of extra work to keep track of the fact that all of our chains have "`[0:0]`" for their policy counters.

(The bug that resets the counters with iptables-legacy might be something that annoys some user somewhere, but AFAICT no one ever filed a bug report about it...)

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1839,27 +1839,16 @@ func (proxier *Proxier) acceptIPVSTraffic() {
 
 // createAndLinkKubeChain create all kube chains that ipvs proxier need and write basic link.
 func (proxier *Proxier) createAndLinkKubeChain() {
-	existingFilterChains := proxier.getExistingChains(proxier.filterChainsData, utiliptables.TableFilter)
-	existingNATChains := proxier.getExistingChains(proxier.iptablesData, utiliptables.TableNAT)
-
 	// Make sure we keep stats for the top-level chains
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1842]: this comment no longer applies

File: pkg/util/iptables/save_restore.go [hunk 2]
```diff
@@ -21,102 +21,32 @@ import (
...
-	readIndex := 0
-	// find beginning of table
-	for readIndex < len(save) {
-		line, n := readLine(readIndex, save)
-		readIndex = n
-		if bytes.HasPrefix(line, tablePrefix) {
+// GetChains parses iptables-save data to find the chains that are defined.
+// It assumes that save contains a single table's data, and returns a map with
+// keys for every chain defined in that table.
+func GetChains(save []byte) map[Chain]struct{} {
```
Comment (aojea) [inline pkg/util/iptables/save_restore.go:32]: GetChainsFromTable ??

File: pkg/util/iptables/save_restore.go [hunk 3]
```diff
@@ -21,102 +21,32 @@ import (
...
+	chainsMap := make(map[Chain]struct{})
+
+	for {
+		i := bytes.Index(save, []byte("\n:"))
```
Comment (aojea) [inline pkg/util/iptables/save_restore.go:36]: the logic is much simpler now, just get everything between `\n:`  and a blank space

**Comment by aojea** [comment]
LGTM
just the name of the function can be changed to indicate more clear the fact it assumes Tables only
/assign @thockin

**Comment by aojea** [comment]
@danwinship seems we can stop passing ` utiliptables.RestoreCounters` to the restore

https://github.com/kubernetes/kubernetes/blob/a3da8d73004a8727574d96ebb38dbbe719c6b3ab/pkg/proxy/iptables/proxier.go#L1504

**Comment by danwinship** [comment]
yeah... I didn't bother because it doesn't simplify the code any (it makes it two characters longer!)

I guess we could just remove the argument from the function entirely, but...

**Comment by aojea** [comment]
> yeah... I didn't bother because it doesn't simplify the code any (it makes it two characters longer!)
> 
> I guess we could just remove the argument from the function entirely, but...

nah, I just found it , it can be a follow up

**Pull request merged**